src/commands/locks/common.cc

```diff
@@ -0,0 +1,100 @@
+/*---------------------------------------------------------------------------*\
+                          ____  _ _ __ _ __  ___ _ _
+                         |_ / || | '_ \ '_ \/ -_) '_|
+                         /__|\_, | .__/ .__/\___|_|
+                             |__/|_|  |_|
+\*---------------------------------------------------------------------------*/
+/** \file commands/locks/common.cc
+ * Common code used by different commands.
+ */
+#include "commands/locks/common.h"
+#include "repos.h"
+
+// OLD STYLE VERSIONED LOCKS:
+//	solvable_name: kernel
+//	version: > 1
+//
+// NEW STYLE VERSIONED LOCKS:
+//	complex: AttrMatchData solvable:name kernel C SolvableRange\ >\ 1\ \"\"
+//   or
+//	solvable_name: kernel > 1
+//
+// Semantically equivalent as locks, but due to the different syntax
+// the complex lock is wrongly handled in list. Different syntax also
+// may prevent removing locks (old and new style locks are not ==).
+//
+// bsc#1112911: Unfortunately all styles are found in real-life locks-files.
+// libzypp will try to make sure, when parsing the locks-file, that only
+// OLD STYLE queries are generated. They should work for list and remove.
+#undef	ENABLE_NEW_STYLE_VERSIONED_LOCKS
+
+///////////////////////////////////////////////////////////////////
+namespace locks
+{
+  using namespace zypp;
+  ///////////////////////////////////////////////////////////////////
+  namespace
+  {
+    inline void addNameDependency( PoolQuery & q_r, const std::string & arg_r )
+    {
+      CapDetail d { Capability(arg_r) };
+      if ( d.isVersioned() )
+      {
+#ifdef ENABLE_NEW_STYLE_VERSIONED_LOCKS
+	q_r.addDependency( sat::SolvAttr::name, d.name().asString(), d.op(), d.ed() );
+#else
+	q_r.addAttribute( sat::SolvAttr::name, d.name().asString() );
+	q_r.setEdition( d.ed(), d.op() );
+#endif
+      }
+      else
+      {
+#ifdef ENABLE_NEW_STYLE_VERSIONED_LOCKS
+	q_r.addDependency( sat::SolvAttr::name, arg_r );
+#else
+	q_r.addAttribute( sat::SolvAttr::name, arg_r );
+#endif
+      }
+    }
+  }
+  ///////////////////////////////////////////////////////////////////
+
+  PoolQuery arg2query( Zypper & zypper, const std::string & arg_r, const std::set<ResKind> & kinds_r, const std::vector<std::string> & repos_r )
+  {
+    // Try to stay with the syntax the serialized query (AKA lock) generates.
+    //     type: package
+    //     match_type: glob
+    //     case_sensitive: on
+    //     solvable_name: kernel
+
+    PoolQuery q;
+    q.setMatchGlob();
+    q.setCaseSensitive();
+
+    for_( it, repos_r.begin(), repos_r.end() )
+    {
+      RepoInfo info;
+      if ( match_repo( zypper, *it, &info ) )
+	q.addRepo( info.alias() );
+      else //TODO some error handling
+	WAR << "unknown repository" << *it << endl;
+    }
+
+    if ( kinds_r.empty() ) // derive it from the name
```

Comment:
This should be changed. If the arg contains an explicit kind like `pattern:foo`, then the query should also be built by the `if`. 
It makes no sense to add all kinds from the command line, but then overwrite this by adding an explicit search for `pattern:foo`.  A simple `kind=pattern` `name =foo` is enough.

Reply:
ah right, thats broken otherwise